openapi-model/src/main/scala/sttp/apispec/openapi/validation/OpenAPIComparator.scala

```diff
@@ -0,0 +1,277 @@
+      None
+    else
+      Some(IncompatibleContent(issues.toList))
+  }
+
+  private def checkMediaType(
+      mediaType: String,
+      clientMediaTypeDescription: MediaType,
+      serverMediaTypeDescription: MediaType
+  ): Option[IncompatibleMediaType] = {
+    val issues = checkSchema(clientMediaTypeDescription.schema, serverMediaTypeDescription.schema)
+    if (issues.nonEmpty)
+      Some(IncompatibleMediaType(mediaType, issues.toList))
+    else
+      None
+    // TODO: encoding?
+  }
+
+  private def checkSchema(
+      clientSchema: Option[SchemaLike],
+      serverSchema: Option[SchemaLike]
+  ): Option[OpenAPICompatibilityIssue] = {
+    (serverSchema, clientSchema) match {
+      case (Some(serverSchema), Some(clientSchema)) =>
+        val schemaComparator = new SchemaComparator(clientSchemas, serverSchemas)
+        val schemaIssues = schemaComparator.compare(clientSchema, serverSchema)
+        if (schemaIssues.nonEmpty)
+          Some(IncompatibleSchema(schemaIssues))
+        else
+          None
+      case _ => None
+
+    }
+  }
+
+  private def getOperationParameters(operation: Operation): List[Parameter] = {
+    operation.parameters.flatMap {
+      case Right(parameter) => Some(parameter)
+      case Left(reference)  => resolveParameterReference(serverOpenAPI, reference.$ref)
+    }
+  }
+
+  private def resolveParameterReference(openAPI: OpenAPI, ref: String): Option[Parameter] = {
+    openAPI.components match {
+      case Some(component) => component.getLocalParameter(ref)
+      case None            => None
+    }
+  }
+
+  private def checkRequestBody(
+      clientRequestBody: RequestBody,
+      serverRequestBody: RequestBody
+  ): Option[IncompatibleRequestBody] = {
+    val contentIssues = checkContent(clientRequestBody.content, serverRequestBody.content).toList
+    if (contentIssues.nonEmpty)
+      Some(IncompatibleRequestBody(contentIssues))
+    else
+      None
+  }
+
+  private def checkResponse(clientResponse: Response, serverResponse: Response): Option[IncompatibleResponse] = {
+    val contentIssue = checkContent(clientResponse.content, serverResponse.content)
+    val headerIssues = clientResponse.headers.flatMap {
+      case (clientHeaderName, Right(clientHeader)) =>
+        val serverHeader = serverResponse.headers.get(clientHeaderName)
+        serverHeader match {
+          case Some(Right(serverHeader)) =>
+            if (clientHeader.required.getOrElse(false) && !serverHeader.required.getOrElse(false)) {
+              Some(IncompatibleRequiredHeader(clientHeaderName))
+            } else {
+              checkHeader(clientHeaderName, clientHeader, serverHeader)
+            }
+          case None => Some(MissingHeader(clientHeaderName))
+          case _    => None
+        }
+      case _ => None
+    }
+
+    val issues = contentIssue.toList ++ headerIssues
+    if (issues.nonEmpty)
+      Some(IncompatibleResponse(issues))
+    else
+      None
+  }
+
+  private def checkHeader(
```

Comment:
Headers can appear both in requests and responses, but the `Header` class is only used for responses. Just for clarity, it would be nice to rename this method to something like `checkResponseHeader`